cilium/conntrack.cc

```diff
@@ -0,0 +1,282 @@
+    // do not fail if non-TCP map can not be opened
+  }
+
+  ok_ = true;
+}
+
+CtMap::CtMaps6::CtMaps6(const std::string& bpf_root, const std::string& map_name)
+  : ok_(false) {
+  // Open the IPv6 bpf maps from Cilium specific paths
+
+  std::string path6tcp(bpf_root + "/tc/globals/cilium_ct6_" + map_name);
+  if (!ctmap6_tcp_.open(path6tcp)) {
+    ENVOY_LOG(info, "cilium.bpf_metadata: Cannot open IPv6 conntrack map at {}", path6tcp);
+    return;
+  }
+  std::string path6any(bpf_root + "/tc/globals/cilium_ct_any6_" + map_name);
+  if (!ctmap6_any_.open(path6any)) {
+    ENVOY_LOG(info, "cilium.bpf_metadata: Cannot open IPv6 conntrack map at {}", path6any);
+    // do not fail if non-TCP map can not be opened
+  }
+
+  ok_ = true;
+}
+
+// Must hold mutex!
+std::unordered_map<const std::string, std::unique_ptr<CtMap::CtMaps4>>::iterator
+CtMap::openMap4(const std::string& map_name) {
+  std::string name(map_name);
+  auto pair = ct_maps4_.emplace(std::make_pair(name, nullptr));
+  // construct the maps only if the entry was inserted
+  if (pair.second) {
+    auto maps = new CtMaps4(bpf_root_, name);
+    if (!maps->ok_) {
+      // Map open failed, delete and return nullptr
+      delete maps;
+      ct_maps4_.erase(pair.first);
+      return ct_maps4_.end();
+    }
+    pair.first->second.reset(maps);
+  }
+  ENVOY_LOG(trace, "cilium.bpf_metadata: Opened IPv4 conntrack map {}", name);
+  return pair.first;
+}
+
+// Must hold mutex!
+std::unordered_map<const std::string, std::unique_ptr<CtMap::CtMaps6>>::iterator
+CtMap::openMap6(const std::string& map_name) {
+  std::string name(map_name);
+  auto pair = ct_maps6_.emplace(std::make_pair(name, nullptr));
+  // construct the maps only if the entry was inserted
+  if (pair.second) {
+    auto maps = new CtMaps6(bpf_root_, name);
+    if (!maps->ok_) {
+      // Map open failed, delete and return nullptr
+      delete maps;
+      ct_maps6_.erase(pair.first);
+      return ct_maps6_.end();
+    }
+    pair.first->second.reset(maps);
+  }
+  ENVOY_LOG(trace, "cilium.bpf_metadata: Opened IPv6 conntrack map {}", name);
+  return pair.first;
+}
+
+void CtMap::closeMaps(const std::shared_ptr<std::unordered_set<std::string>>& to_be_closed) {
+  std::lock_guard<std::mutex> guard(maps_mutex_);
+
+  for (const auto& map_name: *to_be_closed) {
+    const std::string& name = (map_name != "") ? map_name : "global";
+    auto ct4 = ct_maps4_.find(name);
+    if (ct4 != ct_maps4_.end()) {
+      ct_maps4_.erase(ct4);
+      ENVOY_LOG(trace, "cilium.bpf_metadata: Closed IPv4 conntrack map {}", name);
+    }
+    auto ct6 = ct_maps6_.find(name);
+    if (ct6 != ct_maps6_.end()) {
+      ct_maps6_.erase(ct6);
+      ENVOY_LOG(trace, "cilium.bpf_metadata: Closed IPv6 conntrack map {}", name);
+    }
+  }
+}
+
+CtMap::CtMap(const std::string &bpf_root) : bpf_root_(bpf_root) {
+  // Try to open the bpf maps from Cilium specific paths
+  openMap4("global");
+  openMap6("global");
```

Comment:
Should we try to error-check here / catch&log failures early? Presumably the CT map should be guaranteed to exist when the proxy is launched?

Reply:
I guess we don't actually know if this is intentional, maybe the user just configured Cilium with IPv4/IPv6 disabled... in which case warning about it wouldn't really be right---info, if anything.

Reply:
+1

If ipv4 or ipv6 are disabled, then specific maps don't even exist.